### PR TITLE
Make SQLDatabaseQueue less susceptible to Race Conditions

### PR DESCRIPTION
**Pull request by rhyshort**
What:

Reduce the possible window for a task to be added to the SQLiteDatabaseQueue after shutdown has been called.

How:

Introduce a flag which enables SQLDatabase Queue to reject tasks, as soon as shutdown is called.
When the the queue can no longer accept jobs, RejectedExecutionException is thrown. During the shutdown method, will wait 2 seconds to allow other jobs to finish.

Things to look at:

I'm not sure of waiting two seconds is appropriate, but having the ExecutorService thread killed before the job is complete is a bit of a risk, any thoughts on this? 

BugzId: 47164

reviewer @tomblench 
reviewer @mikerhodes

### CODE REVIEW

File: sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java [hunk 1]
```diff
@@ -128,6 +135,11 @@ public void run() {
             }
         });
         queue.shutdown();
+        try {
+            queue.awaitTermination(200, TimeUnit.SECONDS);
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: This wait is for 200 seconds but your PR comments say 2 seconds.

File: sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java [hunk 2]
```diff
@@ -42,6 +45,8 @@
      */
     public SQLDatabaseQueue(String filename) throws IOException {
         this.db = SQLDatabaseFactory.createSQLDatabase(filename);
+        //can directly add to queue no need to check
```
Comment (tomblench) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: No need to check what and why can we directly add?

File: sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java (hunk 1 above)
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: Why bother waiting here? I'm not sure what it add to it.

Really I think you want something which calls `shutdownNow()` once `db.close()` has been called. Can you call it from the task itself?

Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: I guess we could call `shutdownNow()` in the task running on the queue,the tests pass with that behavior, at least on JavaSE.

File: sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java (hunk 2 above)
Comment (rhyshort) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: the next line finishes the comment

Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: But still doesn't explain why there's no need to check, and it is ambiguous what the `it` refers to in the following line, the `queue` or the directly added thing (presumably the runnable?).

File: sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java [hunk 3]
```diff
@@ -121,6 +126,8 @@ public Integer call() throws Exception {
      * tasks
      */
     public void shutdown() {
+        acceptTasks = false;
+        //pass straight to queue, tasks passed via submitTaskToQueue will now be blocked.
```
Comment (mikerhodes) [inline sync-core/src/main/java/com/cloudant/sync/sqlite/SQLDatabaseQueue.java]: Not blocked but rejected

**Comment by mikerhodes** [comment]
I think the approach of using the flag to reject new tasks is sound.

What I don't understand is what behaviour you are trying to achieve in `shutdown`? Is it meant to be blocking? Are you aiming that the task shutdown puts on the queue is the final executed task?

**Comment by rhyshort** [comment]
The main aim is shutdown puts the final task on the queue, which is the one which closes the database. Meaning no other task should hit the queue when it is put on. My idea behind waiting for shutdown to finish, is to make the DatabaseClosed event to be fired when the DB is actually closed, not that we've shutdown the DB queue. Does that make sense?

**Comment by rhyshort** [comment]
Actually I realised the way I was doing it was silly. Instead of waiting on the queue to terminate, calling `get` on the future returned from submit, would have a similar effect. But it will be more explicit what is going on

**Comment by mikerhodes** [comment]
Unless you make `acceptTasks` volatile, you don't get a guarantee that your task is last on the queue. However, I'd suggest avoiding doing that because the likelihood of problems arising from this is very small and the performance cost would be high. Therefore you don't offer a solid guarantee.

Await termination was probably the better way to do it, but with a timeout of something like five minutes rather than a very short one.

I _think_ this is right?

**Comment by mikerhodes** [comment]
:+1:

**Comment by tomblench** [comment]
+1

**Pull request merged**